cpp/src/io/csv/datetime.cuh

```diff
@@ -22,23 +22,18 @@
  * This helper function takes a string and a search range to return the location
  * of the first instance of the specified character.
  *
- * @param[in] data Pointer to the data block
- * @param[in] start_idx Starting index within the data block
- * @param[in] end_idx Ending index within the data block
+ * @param[in] begin Beginning of the character string
+ * @param[in] end End of the character string
  * @param[in] c Character to find
  *
  * @return index into the string, or -1 if the character is not found
  */
-__inline__ __device__ long findFirstOccurrence(const char *data,
-                                               long start_idx,
-                                               long end_idx,
-                                               char c)
+#include "thrust/reduce.h"
+__inline__ __device__ char const* findFirstOccurrence(char const* begin, char const* end, char c)
 {
-  for (long i = start_idx; i <= end_idx; ++i) {
-    if (data[i] == c) { return i; }
-  }
+  while (begin < end and *begin != c) { begin++; }
 
-  return -1;
+  return begin;
```

Comment:
This is an odd function. Returns `begin` but also is modifying it - which is concerning because it is pointer parameter passed in.
1. This isn't consistent with what the function did before (although maybe that is intended).
2. If this is intentional - I would add a comment to the doxygen documentation
3. More ideally this would behave more like `std::find_if` which returns an pointer without modifying the input

Reply:
`begin` is passed-by-value. It's being used as a const iterator. I haven't checked the exact semantics, but afaict this is just `std::find`.

Reply:
we're returning begin so we can then check if it's `>= end`, rather than checking against the `-1` sentinel.

Reply:
And thrust has a `find_if` available. So that's nice.

Reply:
Yeah, I was avoiding thrust functions in this pr for simplicity since robust unit tests are still on the todo list. Wanted to keep the logic as 1-to-1 as possible. If you’d rather I use thrust::find, thats cool with me.

Reply:
lol, shows how confused I am by `const char*` that I missed the fact that the pointer is passed by value. Godbolt link to prove it: https://www.godbolt.org/z/fabqf5

Reply:
I'd rather it be `char const *`, myself, but I'm not sure that's any clearer... :grimacing: